lib/kcerts/ssh.go

```diff
@@ -148,10 +149,14 @@ func (a SSHAgent) AddCertificates(privateKey PrivateKey, publicKey ssh.PublicKey
 		return fmt.Errorf("public key is not a valid ssh certificate")
 	}
 	agentClient := agent.NewClient(conn)
+	ttl := SSHCertRemainingTTL(cert).Seconds()
+	if math.Signbit(ttl) {
```

Comment:
uhm .. let's say the ValidBefore is -2**63, if I subtract the current time, it'll underflow and SSHCertRemainigTTL may return a positive value?

I'd personally be much more comfortable with SSHCertRemainingTTL and friend just erroring out directly if the value in the cert is invalid.

Eg,
```
func SSHCertRemainingTTL(...) (time.Time, error) {
  if ...
}
```

Reply:
So it cant underflow, the Wall of time is casted to a uint64.

`	
ts := time.Unix(int64(math.Pow(2, 64)), 0 )
fmt.Println("printing seconds")
fmt.Println("ts is ", ts.String())
fmt.Println(ts.Sub(time.Unix(0, `0)).Seconds())
`
something like this still returns negative
and arbitrarily adding errors to something like ttl I don't like, because the caller decides whether or not it likes the TTL

Reply:
Let me put it a different way: we have a client, that is receiving a certificate from a server.
There's two fields that by spec should only have positive values. Negative values are invalid?
But the code uses those values anyway, doing some math on them.
It's hard for me to tell that the caller can / will detect all bad combinations.
(example: I don't see the uint64 conversion, but do see the result you describe if I run this on play.google.com)

The code in the golang library either rejects or handles correctly invalid values, with very simple code:
https://cs.opensource.google/go/x/crypto/+/a769d52b:ssh/certs.go;l=404

I think we should do the same? If you don't want to return error, return some value indicating that it has expired?

There's also a value that represents infinity: CertTimeInfinity, which sould be handled differently?

Reply:
Added in more checks and fixed some broken tests with auth_server